src/coreComponents/common/Span.hpp

```diff
@@ -0,0 +1,198 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 Total, S.A
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file Span.hpp
+ */
+
+#ifndef GEOSX_CODINGUTILITIES_SPAN_HPP
+#define GEOSX_CODINGUTILITIES_SPAN_HPP
+
+#include "codingUtilities/traits.hpp"
+#include "common/Logger.hpp"
+
+#include <memory>
+#include <iterator>
+
+namespace geosx
+{
+
+/**
+ * @brief Lightweight non-owning wrapper over a contiguous range of elements.
+ * @tparam T type of range element
+ *
+ * This is a simplified version of C++17's std::span<> that doesn't support
+ * compile-time extents.
+ */
+template< typename T >
+class Span
```

Comment:
Yes, `gsl::span` came to mind, but to be honest, writing a simplified implementation myself probably took me less time than: a) deciding which implementation to include, b) deciding on submodule vs TPL route, c) making sure there are no compilation issues or conflicts (for example, their page clams Clang 10 and 11 supported, but we're building with Clang 8). I don't like reinventing the bicycle as much as anyone, but this is one case where I think it's fine (of course, the caveat is I haven't written any test for it, which will take way more time than the class itself). I don't have strong feelings about it though, I'll happily replace it with something else if everyone agrees it's a better way.

As for `std::span`, it should be more or less a straightforward find/replace, because I kept mine more or less a subset of `std::span`'s interface (albeit, c++20 version), with fewer constructors and lacking a few methods and a defaulted template parameter. Speaking of which, I'd like us to have a conversation about moving to C++17 soon :)

Reply:
That's also the opportunity to discuss how we could do this kind of inclusions for other cases that are header only.
For example, I was using https://github.com/nlohmann/json. Just
```cmake
include(FetchContent)

FetchContent_Declare(json
  GIT_REPOSITORY https://github.com/nlohmann/json.git
  GIT_TAG v3.7.3)

FetchContent_GetProperties(json)
if(NOT json_POPULATED)
  FetchContent_Populate(json)
  add_subdirectory(${json_SOURCE_DIR} ${json_BINARY_DIR} EXCLUDE_FROM_ALL)
endif()

target_link_libraries(foo PRIVATE nlohmann_json::nlohmann_json)
```
and you're done, that's cool.

> Speaking of which, I'd like us to have a conversation about moving to C++17 soon :)

Save your breath.

C++ 17 will surely [require CUDA 11+](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#c-cplusplus-language-support) which on [Power requires RHEL 8](https://developer.nvidia.com/cuda-11.0-download-archive?target_os=Linux&target_arch=ppc64le&target_distro=RHEL&target_version=8) where Pangea3 is RHEL 7.
And I have no insight about any update.
Furthermore, CUDA 10 does not accept g++ versions greater than 8 (if I remember correctly), so we cannot even try to trick some features in... (which may be better 🤣 )

And that's so sad to be kept with old C++ 😢